src/corelibs/U2View/src/ov_msa/exclude_list/MsaExcludeList.cpp

```diff
@@ -0,0 +1,593 @@
+    connect(moveToMsaAction, &QAction::triggered, this, &MsaExcludeListWidget::moveExcludeListSelectionToMaObject);
+
+    auto moveToMsaButton = new QToolButton();
+    moveToMsaButton->setObjectName("exclude_list_move_to_msa_button");
+    moveToMsaButton->setDefaultAction(moveToMsaAction);
+    moveToMsaButton->setToolButtonStyle(Qt::ToolButtonIconOnly);
+    toolbarLayout->addWidget(moveToMsaButton);
+    toolbarLayout->addStrut(10);
+
+    auto moveFromMsaAction = viewContext->getMoveMsaSelectionToExcludeListAction(msaEditor);
+    if (moveFromMsaAction != nullptr) {
+        auto moveFromMsaButton = new QToolButton();
+        moveFromMsaButton->setObjectName("exclude_list_move_to_exclude_list_button");
+        moveFromMsaButton->setDefaultAction(moveFromMsaAction);
+        moveFromMsaButton->setToolButtonStyle(Qt::ToolButtonIconOnly);
+        toolbarLayout->addWidget(moveFromMsaButton);
+        toolbarLayout->addStrut(10);
+    }
+
+    auto msaObject = msaEditor->getMaObject();
+    connect(msaObject, &MultipleSequenceAlignmentObject::si_alignmentChanged, this, &MsaExcludeListWidget::handleUndoRedoInMsaEditor);
+
+    excludeListFilePath = msaEditor->property(PROPERTY_LAST_USED_EXCLUDE_LIST_FILE).toString();
+    if (excludeListFilePath.isEmpty() || !QFileInfo::exists(excludeListFilePath)) {
+        GUrl msaUrl = msaObject->getDocument()->getURL();
+        excludeListFilePath = msaUrl.dirPath() + "/" + msaUrl.baseFileName() + "." + EXCLUDE_LIST_FILE_SUFFIX;
+        if (GUrl(excludeListFilePath) == msaUrl) {
+            excludeListFilePath = msaUrl.getURLString() + "." + EXCLUDE_LIST_FILE_SUFFIX;
+        }
+    }
+
+    connect(AppContext::getTaskScheduler(), &TaskScheduler::si_stateChanged, this, &MsaExcludeListWidget::trackMsaObjectSaveTask);
+
+    toolbarLayout->addStretch();
+
+    loadExcludeList(true);
+}
+
+MsaExcludeListWidget::~MsaExcludeListWidget() {
+    if (isDirty) {
+        runSaveTask(excludeListFilePath);
+    }
+}
+
+int MsaExcludeListWidget::addEntry(const DNASequence &sequence, int excludeListRowId) {
+    SAFE_POINT(sequence.alphabet != nullptr, "Sequence must be fully defined!", 0);  // By default MSA row sequences have no alphabet. Catch this kind of error ASAP.
+    int computedExcludeListRowId = excludeListRowId <= 0 ? ++excludeListRowIdGenerator : excludeListRowId;
+    auto item = new QListWidgetItem();
+    item->setText(sequence.getName());
+    item->setData(LIST_ITEM_DATA_ROW_ID, computedExcludeListRowId);
+    nameListView->addItem(item);
+    sequenceByExcludeListRowId[computedExcludeListRowId] = sequence;
+    isDirty = true;
+    return computedExcludeListRowId;
+}
+
+int MsaExcludeListWidget::addMsaRowEntry(const MultipleAlignmentRow &row, int excludeListRowId) {
+    DNASequence sequence = row->getUngappedSequence();
+    sequence.alphabet = msaEditor->getMaObject()->getAlphabet();  // TODO: MSA row must return a fully defined sequence with a correct alphabet.
```

Comment:
The sequence alphabet could be differ from the alignment alphabet, isn't it?

Reply:
I only documented the problem and will work on a solution in a separate PR.
The bug is that any sequence in the DB must have a valid alphabet. 

It is perfectly OK if a sequence has an alphabet compatible with an alignment and is not OK otherwise.